pkg/kudoctl/cmd/diagnostics/builder.go

```diff
@@ -0,0 +1,77 @@
+package diagnostics
+
+import (
+	"github.com/spf13/afero"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// ResourceFnWithContext - resource provider that can read or modify processingContext
+type ResourceFnWithContext func(*ResourceFuncsConfig, *processingContext) (runtime.Object, error)
+type ResourceFn func(*ResourceFuncsConfig) (runtime.Object, error)
+
+func (fn ResourceFnWithContext) toResourceFn(ctx *processingContext) ResourceFn {
+	return func(r *ResourceFuncsConfig) (runtime.Object, error) {
+		return fn(r, ctx)
+	}
+}
+
+type Builder struct {
+	r          *ResourceFuncsConfig
+	ctx        processingContext
+	collectors []collector
+	fs         afero.Fs
+}
+
+func (b *Builder) AddGroup(fns ...func(*Builder) collector) *Builder {
+	if len(fns) == 0 {
+		return b
+	}
+	var collectors []collector
+	for _, fn := range fns {
+		collectors = append(collectors, fn(b))
+	}
+	b.collectors = append(b.collectors, &compositeCollector{collectors: collectors})
+	return b
+}
+
+func (b *Builder) Add(fn func(*Builder) collector) *Builder {
```

Comment:
Nit: I see that this isn't the same as calling `AddGroup` with a single element because it's not appending a `compositeCollector`. But, IMO, it could be mistaken as such. How about changing `AddGroup` to

```go
func (b *Builder) AddGroup(fns []func(*Builder) collector) *Builder {
```
which makes the intent a bit clearer. Though, feel free to keep it as-is and just add some documetation for these functions.

Reply:
Request is answered to some extent:
https://github.com/kudobuilder/kudo/blob/ve/simplified-diagnostics/pkg/kudoctl/cmd/diagnostics/runner_helper.go#L29